parsl/providers/provider_base.py

```diff
@@ -29,52 +32,54 @@ class JobStatus(object):
     """Encapsulates a job state together with other details, presently a (error) message"""
     SUMMARY_TRUNCATION_THRESHOLD = 2048
 
-    def __init__(self, state: JobState, message: str = None, exit_code: Optional[int] = None,
-                 stdout_path: str = None, stderr_path: str = None):
+    def __init__(self, state: JobState, message: Optional[str] = None, exit_code: Optional[int] = None,
+                 stdout_path: Optional[str] = None, stderr_path: Optional[str] = None):
         self.state = state
         self.message = message
         self.exit_code = exit_code
         self.stdout_path = stdout_path
         self.stderr_path = stderr_path
 
     @property
-    def terminal(self):
+    def terminal(self) -> bool:
         return self.state.terminal
 
     @property
-    def status_name(self):
+    def status_name(self) -> str:
         return self.state.status_name
 
-    def __repr__(self):
+    def __repr__(self) -> str:
         if self.message is not None:
             return "{} ({})".format(self.state, self.message)
         else:
             return "{}".format(self.state)
 
     @property
-    def stdout(self):
+    def stdout(self) -> Optional[str]:
         return self._read_file(self.stdout_path)
 
     @property
-    def stderr(self):
+    def stderr(self) -> Optional[str]:
         return self._read_file(self.stderr_path)
 
-    def _read_file(self, path):
+    def _read_file(self, path: Optional[str]) -> Optional[str]:
+        if path is None:
```

Comment:
This is awkward. Can path be None?

Reply:
it can because JobStatus can be constructed with `None` paths for stderr and stdout - and is so by default.

This test is already present in the similar _read_summary call lower down.

Pretty much all of the `JobStatus` class behaves as if you pass a None path in the constructor, then all of the file related calls will return a None too - eg `stderr()` or `stderr_summary()`.

I'm not trying to move away from that in this PR, but trying to document it in the types and make behaviour more consistent.